microceph/ceph/osd.go

```diff
@@ -69,15 +72,155 @@ func nextOSD(s *state.State) (int64, error) {
+	}
+
+	// Encode as base64, this results in 128 bytes.
+	return []byte(base64.StdEncoding.EncodeToString(key)), nil
+}
+
+// encryptDevice encrypts the given device with the given key.
+func encryptDevice(path string, key []byte) error {
+	// Run the cryptsetup command.
+	cmd := exec.Command(
+		"cryptsetup",
+		"--batch-mode",
+		"--key-file", "-",
+		"luksFormat",
+		path)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return fmt.Errorf("Error in cryptsetup pipe: %s", err)
+	}
+	if _, err = stdin.Write(key); err != nil {
+		return fmt.Errorf("Error writing key to cryptsetup pipe: %s", err)
+	}
+	stdin.Close()
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Failed to luksFormat device: %s, %s, %s", path, err, out)
+	}
+	return nil
+}
+
+// Store the key in the ceph key value store, under a name that derives from the osd id.
+func storeKey(key []byte, osdID int64) error {
+	// Run the ceph config-key set command
+	_, err := processExec.RunCommand("ceph", "config-key", "set", fmt.Sprintf("microceph:osd.%d/key", osdID), string(key))
+	if err != nil {
+		return fmt.Errorf("Failed to store key: %w", err)
+	}
+	return nil
+}
+
+// Open the encrypted device and return its path.
+func openEncryptedDevice(path string, osdID int64, key []byte) (string, error) {
+	// Run the cryptsetup open command, expect key on stdin
+	cmd := exec.Command(
+		"cryptsetup",
+		"--keyfile-size", "128",
+		"--key-file", "-",
+		"luksOpen",
+		path,
+		fmt.Sprintf("luksosd-%d", osdID),
+	)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return "", fmt.Errorf("Error in cryptsetup pipe: %s", err)
+	}
+	if _, err = stdin.Write(key); err != nil {
+		return "", fmt.Errorf("Error writing key to cryptsetup pipe: %s", err)
+	}
+	stdin.Close()
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf(`Failed to luksOpen: %s, %s, %s
+
+NOTE: OSD Encryption requires a snapd >= 2.59
+Verify your version of snapd by running "snap version"
+`, path, err, out)
+	}
+	return fmt.Sprintf("/dev/mapper/luksosd-%d", osdID), nil
+}
+
+// checkEncryptSupport checks if the kernel supports encryption.
+// Checks performed:
+// - Check if the kernel module is loaded.
+// - Check if we have a mapper control file.
+// - Check if we can access /run
+func checkEncryptSupport() error {
+	// Check if we have a mapper
+	if _, err := os.Stat("/dev/mapper/control"); err != nil {
+		return fmt.Errorf("Missing /dev/mapper/control: %w", err)
+	}
+
+	// Check if we have the dm_crypt module
+	inf, err := os.Stat("/sys/module/dm_crypt")
+	if err != nil || inf == nil || !inf.IsDir() {
+		return fmt.Errorf("Missing dm_crypt module: %w", err)
+	}
+
+	// Check if we can list the /run directory; older snapd had an issue with this, https://github.com/snapcore/snapd/pull/12445
+	if _, err = os.ReadDir("/run"); err != nil {
+		return fmt.Errorf("Can't access /run, might need to update snapd to >2.58: %w", err)
```

Comment:
Looks like this should be updated to 2.59, matching the line above

Reply:
Heh, the /run bug was really fixed earlier than the other one above (in 2.58.1 or so) so this line technically is correct. But guess this might be confusing and doesn't really help... I'll push a change